src/server/tiered_storage.cc

```diff
@@ -251,10 +294,60 @@ bool TieredStorage::ShardOpManager::NotifyDelete(tiering::DiskSegment segment) {
   return false;
 }
 
+void TieredStorage::ShardOpManager::RetireColdEntries(size_t additional_memory) {
+  ssize_t threshold = memory_low_limit_ + additional_memory;
+
+  // we trigger if we below the low point
+  if (db_slice_.memory_budget() < threshold) {
+    // But once we below we try to raise above high watermark.
+    const double kHighFactor = 1.0;
+    size_t needed_to_free =
+        (threshold - db_slice_.memory_budget()) + memory_low_limit_ * kHighFactor;
+    size_t gained = 0;
+    do {
+      size_t memory_before = ts_->cool_queue_.UsedMemory();
+      detail::TieredColdRecord* record = ts_->cool_queue_.PopBack();
+      if (record == nullptr)  // nothing to pull anymore
+        break;
+
+      auto predicate = [record](const PrimeKey& key, const PrimeValue& probe) {
+        return probe.IsExternal() && probe.IsCool() && probe.GetCool().record == record;
+      };
+
+      PrimeIterator it =
+          db_slice_.GetDBTable(record->db_index)->prime.FindFirst(record->key_hash, predicate);
+      CHECK(IsValid(it));
+      PrimeValue& pv = it->second;
+      tiering::DiskSegment segment = FromCoolItem(pv.GetCool());
+
+      // Now the item is only in storage.
+      pv.SetExternal(segment.offset, segment.length);
+      CompactObj::DeleteMR<detail::TieredColdRecord>(record);
+      gained += memory_before - ts_->cool_queue_.UsedMemory();
```

Comment:
I think you forgot to update the type stats

Reply:
true